File: src/Web/Dart/observer.dart
typedef Observer = String Function(int id);

final class Subject {
  final Map<String, Observer> _observers = <String, Observer>{};

  bool subscribe(String key, Observer observer) {
    if (_observers.containsKey(key)) return false;
    _observers[key] = observer;
    return true;
  }

  bool unsubscribe(String key) => _observers.remove(key) != null;

  List<String> publish(int id) => <String>[
    for (final observer in _observers.values) observer(id),
  ];
}

bool observerExamplePasses() {
  final subject = Subject();
  final auditAdded = subject.subscribe('audit', (id) => 'audit:$id');
  final dashboardAdded = subject.subscribe(
    'dashboard',
    (id) => 'dashboard:$id',
  );
  final duplicateRejected = !subject.subscribe(
    'audit',
    (id) => 'duplicate:$id',
  );

  final first = subject.publish(42).join('>');
  final dashboardRemoved = subject.unsubscribe('dashboard');
  final second = subject.publish(43).join('>');

  return auditAdded &&
      dashboardAdded &&
      duplicateRejected &&
      first == 'audit:42>dashboard:42' &&
      dashboardRemoved &&
      second == 'audit:43';
}

File: src/Other/MicroPython/observer.py
class RecordingObserver:
    def __init__(self, name):
        self.name = name
        self.events = []

    def update(self, state):
        self.events.append(state)


class Subject:
    def __init__(self):
        self._observers = []

    def subscribe(self, observer):
        if observer in self._observers:
            return False
        self._observers.append(observer)
        return True

    def unsubscribe(self, observer):
        if observer not in self._observers:
            return False
        self._observers.remove(observer)
        return True

    def publish(self, state):
        for observer in tuple(self._observers):
            observer.update(state)


subject = Subject()
audit = RecordingObserver("audit")
dashboard = RecordingObserver("dashboard")

assert subject.subscribe(audit)
assert subject.subscribe(dashboard)
assert not subject.subscribe(audit)

subject.publish("draft")

assert subject.unsubscribe(dashboard)
assert not subject.unsubscribe(dashboard)

subject.publish("published")

assert audit.events == ["draft", "published"]
assert dashboard.events == ["draft"]

print("observer=audit:draft,published;dashboard:draft;duplicate=rejected;second-unsubscribe=rejected")
